### PR TITLE
txn: Check whether the primary matches when handling check_txn_status requests

### LINKED ISSUE #14636: Support verifying whether it's primary when handling check_txn_status
## Development Task

This is part of fixing of https://github.com/pingcap/tidb/issues/42937 . 

It's quite triky to make it works well with both new clients and old clients that doesn't support this behavior. To simplify the problem, a field `verify_is_primary` can be added to `CheckTxnStatusRequest`, and TiKV only do this check when this field is set to true.

### PR DESCRIPTION
**Pull request by MyonKeminta**
### What is changed and how it works?

Issue Number: close #14636
ref https://github.com/pingcap/tidb/issues/42937
Requires https://github.com/pingcap/kvproto/pull/1105
Requires https://github.com/pingcap/kvproto/pull/1107

What's Changed:

```commit-message
Makes TiKV support checking whether the lock is primary when handling check_txn_status.
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

```release-note
None
```

### CODE REVIEW

File: src/storage/txn/actions/check_txn_status.rs [hunk 1]
```diff
@@ -24,7 +24,13 @@ pub fn check_txn_status_lock_exists(
     caller_start_ts: TimeStamp,
     force_sync_commit: bool,
     resolving_pessimistic_lock: bool,
+    verify_is_primary: bool,
 ) -> Result<(TxnStatus, Option<ReleasedLock>)> {
+    if verify_is_primary && !primary_key.is_encoded_from(&lock.primary) {
+        // Return the current lock info to tell the client what the actual primary is.
+        return Err(ErrorInner::KeyIsLocked(lock.into_lock_info(primary_key.into_raw()?)).into());
```
Comment (ekexium) [inline src/storage/txn/actions/check_txn_status.rs:31]: I'm wondering whether the semantics of `KeyIsLocked` is clear and explicit. How about defining a new kind of error?

Comment (MyonKeminta) [reply src/storage/txn/actions/check_txn_status.rs:31]: Ok. Please take a look at this https://github.com/pingcap/kvproto/pull/1107

**Review by ekexium — APPROVED** [review]

**Comment by MyonKeminta** [comment]
@zyguan Do you have time to take a look? thx

**Review by zyguan — APPROVED** [review]

**Review by hicqu — APPROVED** [review]

**Pull request merged**